aws/resource_aws_qldb_stream_test.go

```diff
@@ -0,0 +1,315 @@
+			})
+
+		if err == nil {
+			if len(aws.StringValue(resp.Name)) != 0 && aws.StringValue(resp.Name) == rs.Primary.ID {
+				return fmt.Errorf("QLDB Ledger %s still exists", rs.Primary.ID)
+			}
+		}
+
+		// Return nil if the cluster is already destroyed
+		if isAWSErr(err, qldb.ErrCodeResourceNotFoundException, "") {
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckAWSQLDBStreamExists(n string, v *qldb.DescribeJournalKinesisStreamOutput) resource.TestCheckFunc {
+	log.Printf("Checking for QLDB stream's existence... %s", n)
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			log.Printf("Not found: %s", n)
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			log.Printf("No QLDB Stream ID is set")
+			return fmt.Errorf("No QLDB Stream ID is set")
+		}
+
+		ledgerName, ok := rs.Primary.Attributes["ledger_name"]
+		if !ok {
+			log.Printf("No ledger name value has been set")
+			return fmt.Errorf("No ledger name value has been set")
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).qldbconn
+		resp, err := conn.DescribeJournalKinesisStream(&qldb.DescribeJournalKinesisStreamInput{
+			LedgerName: aws.String(ledgerName),
+			StreamId:   aws.String(rs.Primary.ID),
+		})
+
+		if err != nil {
+			log.Printf("Error: %s", err.Error())
+			return err
+		}
+
+		if *resp.Stream.StreamId == rs.Primary.ID {
+			*v = *resp
+			return nil
+		}
+
+		log.Printf("QLDB Stream (%s) not found", rs.Primary.ID)
+		return fmt.Errorf("QLDB Stream (%s) not found", rs.Primary.ID)
+	}
+}
+
+func testAccAWSQLDBStreamDependenciesConfig(rLedgerName, rKinesisStreamName, rRoleName, rStreamName string) string {
+	return fmt.Sprintf(`
+resource "aws_qldb_ledger" "test" {
+	name                = "%s"
+	deletion_protection = false
+}
+
+resource "aws_kinesis_stream" "test" {
+	name             = "%s"
+	shard_count      = 1
+	retention_period = 24
+}
+
+resource "aws_iam_role" "test" {
+	name = "%s"
+
+	assume_role_policy = jsonencode({
+		Version = "2012-10-17"
+		Statement = [
+			{
+				Action = "sts:AssumeRole"
+				Effect = "Allow"
+				Sid    = ""
+				Principal = {
+				Service = "qldb.amazonaws.com"
+				}
+			},
+		]
+	})
+
+	inline_policy {
+		name = "test-qldb-policy"
+		policy = jsonencode({
+		Version = "2012-10-17"
+		Statement = [
+			{
+				Action = [
+					"kinesis:PutRecord*",
+					"kinesis:DescribeStream",
+					"kinesis:ListShards",
+				]
+				Effect   = "Allow"
+				Resource = aws_kinesis_stream.test.arn
+			},
+		]
+		})
+	}
+}
+
+resource "aws_qldb_stream" "test" {
+	stream_name          = "%s"
+	ledger_name          = aws_qldb_ledger.test.id
+	inclusive_start_time = "2021-01-01T00:00:00Z"
+
+	role_arn = aws_iam_role.test.arn
+
+	kinesis_configuration = {
+		aggregation_enabled = false
+		stream_arn          = aws_kinesis_stream.test.arn
+	}
+
+	provisioner "local-exec" {
+		command = "sleep 8"
+	}
```

Comment:
I need some insight into this.

There are timing issues where the `aws_iam_role` is "created" but its policies/permissions aren't in effect when the `aws_qldb_stream` is attempted to be created.  Here's the specific error message:
```
Message_: "QLDB is not authorized to perform action: sts:AssumeRole on provided role: arn:aws:iam::xxxxxxxxxxxx:role/tf-test-role",
```

Reply:
This was caused by a delay in the policy taking effect.  Fixed this with resource dependency to impose an arbitrary delay.  Open to any other ideas that are more explicit.